smtp.go

```diff
@@ -2,209 +2,391 @@ package maddy
-				return "", nil, nil, err
-			}
-
-			results, err := verifyDKIM(bytes.NewReader(b.Bytes()))
+func (endp *SMTPEndpoint) setDefaultPipeline(localDeliveryName, remoteDeliveryName string, localOpts, remoteOpts map[string]string) error {
+	var err error
+	var localDelivery module.DeliveryTarget
+	if localDeliveryName == "" {
+		localDelivery, err = deliveryTarget("default_local_delivery")
+		if err != nil {
+			localDelivery, err = deliveryTarget("default")
 			if err != nil {
-				return "", nil, nil, err
+				return err
 			}
+		}
+		localOpts = map[string]string{"local_only": ""}
+	} else {
+		localDelivery, err = deliveryTarget(localDeliveryName)
+		if err != nil {
+			return err
+		}
+	}
 
-			// TODO: strip existing Authentication-Results header fields with our identity
+	if endp.submission {
+		if remoteDeliveryName == "" {
+			remoteDeliveryName = "default_remote_delivery"
+			remoteOpts = map[string]string{"remote_only": ""}
+		}
+		remoteDelivery, err := deliveryTarget(remoteDeliveryName)
+		if err != nil {
+			return err
+		}
 
-			// TODO: don't print Authentication-Results on a single line
-			authRes := "Authentication-Results: " + msgauth.Format(identity, results) + "\r\n"
-			r = io.MultiReader(strings.NewReader(authRes), &b)
-			return from, to, r, nil
-		},
+		endp.pipeline = append(endp.pipeline,
+			// require_auth and submission_check are always prepended to pipeline
+			//TODO: DKIM sign
+			deliveryStep{t: localDelivery, opts: localOpts},
+			deliveryStep{t: remoteDelivery, opts: remoteOpts},
+		)
+	} else {
+		endp.pipeline = append(endp.pipeline,
+			//TODO: DKIM verify
+			deliveryStep{t: localDelivery, opts: localOpts},
+		)
 	}
+
+	return nil
+}
+
+func (endp *SMTPEndpoint) Login(state *smtp.ConnectionState, username, password string) (smtp.User, error) {
+	if !endp.Auth.CheckPlain(username, password) {
+		return nil, errors.New("Invalid credentials")
+	}
+
+	return SMTPUser{
+		endp:     endp,
+		username: username,
+		state:    state,
+	}, nil
+}
+
+func (endp *SMTPEndpoint) AnonymousLogin(state *smtp.ConnectionState) (smtp.User, error) {
+	return SMTPUser{
+		endp:      endp,
+		anonymous: true,
+		state:     state,
+	}, nil
 }
 
-func newRelay(be smtp.Backend) smtp.Backend {
-	hostname := "localhost" // TODO
+func formatAddr(raw string) string {
+	addr := mail.Address{Address: raw}
+	return addr.String()
+}
 
-	return &backendutil.TransformBackend{
-		Backend: be,
-		Transform: func(from string, to []string, r io.Reader) (string, []string, io.Reader, error) {
-			// TODO: don't print Received on a single line
-			received := "Received:"
-			// TODO: from
-			received += " by " + hostname
-			received += " with ESMTP"
-			// TODO: for
-			received += "; " + time.Now().Format(time.RFC1123Z) + "\r\n"
-			// TODO: add comments with TLS information
+func addReturnPath(ctx module.DeliveryContext, msg io.Reader) io.Reader {
+	hdr := "Return-Path: " + formatAddr(ctx.From) + "\r\n"
```

Comment:
Do we sanitize this to make sure it's doesn't contain `\n`?

(The SMTP server might parse `\n` as not-a-newline since the RFC requires `\r\n`)

Reply:
Totally forgot about this function. I should remove it anyway. And no, we don't sanitize, I assume that go-smtp doesn't return "invalid" values (here: ones with line separators). Will fix.